### DISCUSSION TITLE
useFieldArray map components with own remove method

### DISCUSSION THREAD

**Discussion by tom-dudman** [react-hook-form/react-hook-form#7633] [Q&A] [answered]
I'm trying to render a component for each field from a `useFieldArray`, and give that component a method to remove itself, however I'm seeing some strange behaviour.  Methods I've tried:

1. Finding the field's `index` based on its `id` and remove based on that index, but the field seems to get a different `id` for each `useFieldArray` call
2. Passing the field's index as a prop, and calling remove on that - it doesn't seem to unmount the component, and so the fields get re-registered
3. Passing a specific remove method as a prop to the child - this is the only way I can get it to work

Method 3 is OK, but I'm wondering if there's a way to create an entirely self-contained remove method, passing nothing but the field as a prop.

I've tried to make a super-simple CSB showing all three methods: https://codesandbox.io/s/react-hook-form-field-array-removal-methods-1zg7v

**Comment by Moshyfawn** [comment, marked as answer]
[From RHF docs on `useFieldArray` hook](https://react-hook-form.com/api/usefieldarray)

> useFieldArray automatically generates a unique identifier named id which is used for key prop

> When your array field contains objects with the key name id, useFieldArray will overwrite and remove it. If you want to keep the id field in your array of objects, you must use keyName prop to change to other name

You want to pass the fields array element **index** into the `remove` field array method. 

See [this CSB](https://codesandbox.io/s/mystifying-mcclintock-8seuy?file=/src/App.tsx) for example
